File: benchmarks/micro/AlgebraicIdentitySimplification.c
#include <stdint.h>
#include <unistd.h>

static int64_t normalize(int64_t value) {
    int64_t add = value + 0;
    int64_t multiply = add * 1;
    int64_t masked = multiply & -1;
    int64_t shifted = masked << 0;
    int64_t same_and = shifted & shifted;
    int64_t same_or = same_and | same_and;
    int64_t zero_xor = same_or ^ same_or;
    int64_t zero_and = value & 0;
    int64_t zero_multiply = value * 0;
    int64_t zero_subtract = value - value;
    int64_t all_ones = value | -1;
    return ((shifted ^ 0) + zero_xor + zero_and + zero_multiply + zero_subtract + (all_ones & 1)) - 1;
}

int main(void) {
    int64_t total = (int64_t)getpid();

    for (int32_t i = 0; i < 200000; i += 1) {
        total += normalize(((int64_t)i * 17) + (total % 97));
        total %= 1000000007;
    }

    return total <= 0 ? 1 : 0;
}

File: benchmarks/micro/AlgebraicIdentitySimplification.rs
fn normalize(value: i64) -> i64 {
    let add = value + 0;
    let multiply = add * 1;
    let masked = multiply & -1;
    let shifted = masked << 0;
    let divided = shifted / 1;
    let right_shifted = divided >> 0;
    let same_and = shifted & shifted;
    let same_or = same_and | same_and;
    let zero_xor = same_or ^ same_or;
    let zero_and = value & 0;
    let zero_multiply = value * 0;
    let zero_subtract = value - value;
    let zero_modulo = value % 1;
    let all_ones = value | -1;
    ((right_shifted ^ 0) + zero_xor + zero_and + zero_multiply + zero_subtract + zero_modulo + (all_ones & 1)) - 1
}

fn normalize_slot(slot: i32, salt: i64) -> i64 {
    let modulo = slot % 8;
    let divided = slot / 8;
    salt + modulo as i64 + divided as i64
}

fn normalize_nonzero_slot(slot: i32, salt: i64) -> i64 {
    let divided = slot / slot;
    let modulo = slot % slot;
    salt + divided as i64 + modulo as i64
}

fn normalize_comparisons(value: i32, salt: i64) -> i64 {
    let copy = value;
    let equal = if value == copy { 1 } else { 100 };
    let not_equal = if value != copy { 100 } else { 1 };
    let less = if value < copy { 100 } else { 1 };
    let less_or_equal = if value <= copy { 1 } else { 100 };
    let greater = if value > copy { 100 } else { 1 };
    let greater_or_equal = if value >= copy { 1 } else { 100 };
    salt + (equal + not_equal + less + less_or_equal + greater + greater_or_equal) as i64
}

fn main() {
    let mut total = std::process::id() as i64;

    for i in 0_i32..200_000 {
        total += normalize(((i as i64) * 17) + (total % 97));
        total += normalize_slot(i & 7, total % 41);
        total += normalize_nonzero_slot((i & 7) + 1, total % 43);
        total += normalize_comparisons(i, total % 47);
        total %= 1_000_000_007;
    }

    if total <= 0 {
        std::process::exit(1);
    }
}
